onnxscript/rewriter/fuse_pad_into_conv.py

```diff
@@ -0,0 +1,296 @@
+    """Interface to normalize pad attributes in conv nodes."""
+
+    @staticmethod
+    def compute_pads(
+        input_shape: Sequence[int],
+        output_shape: Sequence[int],
+        attributes: dict[str, Sequence[int] | str],
+    ) -> Sequence[int]:
+        raise NotImplementedError("Child have to implement this function")
+
+    def rewrite(self, op: ir.tape.Tape, conv: ir.Value, **__) -> ir.Value:
+        conv_node = conv.producer()
+
+        # Read spatial dimensions and attributes
+        input_shape = conv_node.inputs[0].shape[2:]
+        output_shape = conv_node.outputs[0].shape[2:]
+        attributes = read_conv_attributes(conv_node)
+
+        # Convert auto_pad mode into an explicit list
+        pads = self.compute_pads(input_shape, output_shape, attributes)
+
+        # Replace auto_pad, forcing to the explicit list
+        conv_attr = conv_node.attributes.copy()
+        conv_attr.add(ir.AttrString("auto_pad", "NOTSET"))
+        if any(x != 0 for x in pads):
+            conv_attr.add(ir.AttrInt64s("pads", pads))
+
+        return op.op(
+            conv_node.op_type,
+            inputs=conv_node.inputs,
+            attributes=conv_attr,
+            domain=conv_node.domain,
+            name=conv_node.name,
+        )
+
+    def check(self, context, conv: ir.Value, **__) -> orp.MatchResult:
+        del context
+        check_result = orp.MatchResult()
+
+        # Conv constraints: attributes
+        conv_node = conv.producer()
+        auto_pad = conv_node.attributes.get_string("auto_pad", None)
+        if auto_pad in [None, "NOTSET"]:
+            return check_result.fail(
+                f"{conv_node.name} auto_pad must be different to 'NOTSET'."
+            )
+
+        # Conv constraints: inputs/outputs
+        input_shape = conv_node.inputs[0].shape
+        output_shape = conv_node.outputs[0].shape
+        if len(input_shape) <= 2:
+            return check_result.fail(f"Input shapes are not defined on {conv_node.name}.")
+        if len(output_shape) <= 2:
+            return check_result.fail(f"Output shapes are not defined on {conv_node.name}.")
+
+        # Conv constraints: values
+        if auto_pad != "VALID":
+            error_msg = "Expected static spatial {} shapes on " + conv_node.name + "."
+            if not all(isinstance(x, int) for x in input_shape[2:]):
+                return check_result.fail(error_msg.format("input"))
+            if not all(isinstance(x, int) for x in output_shape[2:]):
+                return check_result.fail(error_msg.format("output"))
+            attributes = read_conv_attributes(conv_node)
+            if len(attributes["kernel_shape"]) != len(attributes["strides"]):
+                return check_result.fail(
+                    f"strides must have the same length than kernel_shape on {conv_node.name}."
+                )
+        return check_result
+
+
+class NormalizePadFormatConv(_NormalizePadFormatBase):
```

Comment:
I suppose this does not fallback when the rewritten conv still does not match `FusePad` rules? Do we still benefit from converting auto_pad attribute into 'NOTSET' if it's still a no match?

Reply:
You are right, always `NormalizePadFormat` is applied, even if `FusePad` is not applied.

On the other hand, I consider that having a single pad format is coherent, since it makes it easier to be interpreted by different accelerators. ([bug example in `onnxruntime` ](https://github.com/microsoft/onnxruntime/issues/495) due to the `auto_pad` type).

Let me know if you still disagree with this information. What do you think @justinchuby.

Reply:
cc @gramalingam